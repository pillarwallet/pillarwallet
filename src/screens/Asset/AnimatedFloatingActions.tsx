// @flow
/*
    Pillar Wallet: the personal data locker
    Copyright (C) 2019 Stiftung Pillar Project

    This program is free software; you can redistribute it and/or modify
    it under the terms of the GNU General Public License as published by
    the Free Software Foundation; either version 2 of the License, or
    (at your option) any later version.

    This program is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU General Public License for more details.

    You should have received a copy of the GNU General Public License along
    with this program; if not, write to the Free Software Foundation, Inc.,
    51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
*/
import React, { useRef, useState } from 'react';
import t from 'translations/translate';
import { useNavigation } from '@react-navigation/native';
import { Animated, StyleSheet, View } from 'react-native';
import InAppBrowser from 'react-native-inappbrowser-reborn';
import { useIsDarkTheme } from 'utils/themes';
import type { NativeStackNavigationProp as NavigationScreenProp } from '@react-navigation/native-stack';

// Constants
import { BRIDGE_TAB, BRIDGE_FLOW, SEND_TOKEN_FROM_ASSET_FLOW } from 'constants/navigationConstants';
// import { REMOTE_CONFIG } from 'constants/remoteConfigConstants';

// Selectors
import { useIsExchangeAvailable } from 'selectors';

// Components
import FloatingButtons from 'components/FloatingButtons';

// Services
// import { firebaseRemoteConfig } from 'services/firebase';

// Utils
import { pelerinWidgetUrl } from 'utils/fiatToCrypto';
import { showServiceLaunchErrorToast } from 'utils/inAppBrowser';

const AnimatedFloatingActions = ({ assetData }) => {
  const navigation: NavigationScreenProp<any> = useNavigation();
  const isDarkTheme = useIsDarkTheme();

  const isExchangeAvailable = useIsExchangeAvailable();
  const [disableFront, setDisableFront] = useState(false);

  const { contractAddress, token, chain } = assetData;

  // const accounts = useRootSelector(accountsSelector);
  // const activeAccount = getActiveAccount(accounts);
  // const supportedAssets = useSupportedAssetsPerChain();

  const flipAnimation = useRef(new Animated.Value(0)).current;
  let flipRotation = 0;
  flipAnimation.addListener(({ value }) => (flipRotation = value));
  const flipToFrontStyle = {
    transform: [
      {
        rotateY: flipAnimation.interpolate({
          inputRange: [0, 180],
          outputRange: ['0deg', '180deg'],
        }),
      },
    ],
  };
  const flipToBackStyle = {
    transform: [
      {
        rotateY: flipAnimation.interpolate({
          inputRange: [0, 180],
          outputRange: ['180deg', '360deg'],
        }),
      },
    ],
  };

  const flipToFront = () => {
    setTimeout(() => {
      setDisableFront(true);
    }, 500);

    Animated.timing(flipAnimation, {
      toValue: 180,
      duration: 500,
      useNativeDriver: true,
    }).start();
  };
  const flipToBack = () => {
    setDisableFront(false);
    Animated.timing(flipAnimation, {
      toValue: 0,
      duration: 500,
      useNativeDriver: true,
    }).start();
  };

  const sellOnPelerin = async () => {
    const url = await pelerinWidgetUrl(false, null, null, null, 'sell', token, chain);
    await openInAppBrowser(url);
  };

  // const buyOnPelerin = async () => {
  //   const url = await pelerinWidgetUrl(false, null, null, null, 'buy', token, chain);
  //   console.log("urllll", url);
    
  //   await openInAppBrowser(url);
  // };

  // const buyOnRamp = async () => {
  //   const url = onRamperWidgetUrl(activeAccount, supportedAssets);
  //   await openInAppBrowser(url);
  // };

  const openInAppBrowser = async (url) => {
    const isAvailable = await InAppBrowser.isAvailable();

    if (url && isAvailable) {
      InAppBrowser.open(url, {
        // iOS Properties
        dismissButtonStyle: 'close',
        // Android Properties
        showTitle: true,
        enableUrlBarHiding: true,
        enableDefaultShare: true,
      });
    } else showServiceLaunchErrorToast();
  };

  // const mtPelerinSupportedAssetsInString = firebaseRemoteConfig.getString(REMOTE_CONFIG.MT_PELERIN_SUPPORTED_ASSETS);
  
  // const isPelerinSupported = JSON.parse(mtPelerinSupportedAssetsInString).supportedAssets.includes(token);

  // const isRampSupported = ETHERSPOT_RAMP_CURRENCY_TOKENS.includes(token || chain.toUpperCase() + '_' + token);

  const buttons = [
    // (isPelerinSupported || isRampSupported) && {
    //   title: t('label.buy'),
    //   iconName: isDarkTheme ? 'buy' : 'buy-light',
    //   onPress: isPelerinSupported ? buyOnPelerin : buyOnRamp,
    // },
    isExchangeAvailable && {
      title: t('button.swap'),
      iconName: 'exchange',
      onPress: () => {
        if (!!flipRotation) flipToBack();
        else flipToFront();
      },
    },
    {
      title: t('button.send'),
      iconName: 'send',
      onPress: () => navigation.navigate(SEND_TOKEN_FROM_ASSET_FLOW, { assetData }),
    },
    // isPelerinSupported && {
    //   title: t('label.sell'),
    //   iconName: isDarkTheme ? 'sell' : 'sell-light',
    //   onPress: sellOnPelerin,
    // },
  ];

  const exchangeButtons = [
    {
      title: t('label.buy') + ' ' + token,
      iconName: 'arrow-down',
      onPress: () => {
        if (!!flipRotation) flipToBack();
        else flipToFront();
        navigation.navigate(BRIDGE_FLOW, { screen: BRIDGE_TAB, params: { toAssetAddress: contractAddress, chain } });
      },
    },
    {
      title: t('label.sell') + ' ' + token,
      iconName: 'arrow-up',
      onPress: () => {
        if (!!flipRotation) flipToBack();
        else flipToFront();
        navigation.navigate(BRIDGE_FLOW, { screen: BRIDGE_TAB, params: { fromAssetAddress: contractAddress, chain } });
      },
    },
  ];

  return (
    <View style={style.cardWrapper}>
      <Animated.View style={[{ ...style.cardFront, ...flipToBackStyle }, disableFront && { zIndex: 111 }]}>
        <FloatingButtons items={exchangeButtons} />
      </Animated.View>
      <Animated.View style={{ ...style.cardBack, ...flipToFrontStyle }}>
        <FloatingButtons items={!disableFront ? buttons : []} />
      </Animated.View>
    </View>
  );
};

const style = StyleSheet.create({
  cardWrapper: {
    width: '100%',
    alignSelf: 'center',
    position: 'absolute',
    bottom: 0,
  },
  cardFront: {
    minHeight: 100,
    backfaceVisibility: 'hidden',
    width: '100%',
  },
  cardBack: {
    backfaceVisibility: 'hidden',
    minHeight: 100,
    position: 'absolute',
    width: '100%',
  },
});

export default AnimatedFloatingActions;
