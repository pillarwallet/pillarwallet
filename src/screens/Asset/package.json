{
  "name": "Asset",
  "version": "0.0.0",
  "private": true,
  "main": "./Asset.js"
}
