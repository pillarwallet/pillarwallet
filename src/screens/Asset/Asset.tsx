// @flow
/*
    Pillar Wallet: the personal data locker
    Copyright (C) 2019 Stiftung Pillar Project

    This program is free software; you can redistribute it and/or modify
    it under the terms of the GNU General Public License as published by
    the Free Software Foundation; either version 2 of the License, or
    (at your option) any later version.

    This program is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU General Public License for more details.

    You should have received a copy of the GNU General Public License along
    with this program; if not, write to the Free Software Foundation, Inc.,
    51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
*/
import React, { useMemo, useState } from 'react';
import t from 'translations/translate';
import { Dimensions } from 'react-native';
import { useNavigation, useRoute } from '@react-navigation/native';

// Components
import { Container, Content } from 'components/layout/Layout';
import HeaderBlock from 'components/HeaderBlock';
import TokenIcon from 'components/display/TokenIcon';
import { Spacing } from 'components/legacy/Layout';

// Selectors
import { useChainRates, useRootSelector, useActiveAccount } from 'selectors';
import { accountAssetsBalancesSelector } from 'selectors/balances';

// Utils
import { useChainConfig } from 'utils/uiConfig';
import { useTokenDetailsQuery, useMarketDetailsQuery, useHistoricalTokenPriceQuery } from 'utils/etherspot';
import { getAssetRateInFiat } from 'utils/rates';
import { getBalance } from 'utils/assets';
import { convertDecimalNumber } from 'utils/common';
import { isEtherspotAccount } from 'utils/accounts';

// Constants
import { ONE_DAY } from 'constants/assetsConstants';
import { USD } from 'constants/assetsConstants';

// models, types
import type { AssetDataNavigationParam } from 'models/Asset';
import type { Route } from '@react-navigation/native';

// Local
import DurationSelection from './components/DurationSelection';
import YourBalanceContent from './components/YourBalanceContent';
import TokenAnalytics from './components/TokenAnalytics';
import AnimatedGraph from 'components/AnimatedGraph';
import HeaderContent from './components/HeaderContent';
import AnimatedFloatingActions from './AnimatedFloatingActions';

const AssetScreen = () => {
  const navigation = useNavigation();
  const route: Route = useRoute();
  const activeAccount = useActiveAccount();

  const accountAssetsBalances = useRootSelector(accountAssetsBalancesSelector);

  const assetData: AssetDataNavigationParam = route?.params?.assetData;
  const { chain, imageUrl, contractAddress, token } = assetData;

  const { width: screenWidth } = Dimensions.get('window');

  const chainRates = useChainRates(chain);
  const config = useChainConfig(chain);

  const [selectedPeriod, setSelectedPeriod] = useState(ONE_DAY);
  const [pointerTokenValue, setPointerTokenValue] = useState(0);
  const [pointerVisible, setPointerVisible] = useState(false);

  let tokenDetailsQuery = useTokenDetailsQuery(assetData);
  let marketDetailsQuery = useMarketDetailsQuery(assetData);
  let historicalTokenPricesQuery = useHistoricalTokenPriceQuery(assetData, selectedPeriod);

  if (tokenDetailsQuery?.data) {
    const { tokenAddress } = tokenDetailsQuery?.data;
    if (tokenAddress?.toLowerCase() !== contractAddress?.toLowerCase()) {
      tokenDetailsQuery.data = null;
    }
  }
  if (marketDetailsQuery?.data) {
    const { symbol } = marketDetailsQuery?.data;
    if (symbol.toUpperCase() !== token) {
      marketDetailsQuery.data = null;
    }
  }
  if (historicalTokenPricesQuery?.data) {
    const { items } = historicalTokenPricesQuery.data;
    if (items?.[0]?.tokenAddress?.toLowerCase() !== contractAddress?.toLowerCase()) {
      historicalTokenPricesQuery.data = null;
    }
  }

  const networkName = chain ? config.title : undefined;

  const tokenRateCMC = useMemo(() => {
    if (!tokenDetailsQuery.data) return null;
    const { usdPrice } = tokenDetailsQuery.data;
    return convertDecimalNumber(usdPrice);
  }, [tokenDetailsQuery.data]);

  const tokenRate = getAssetRateInFiat(chainRates, contractAddress, USD);
  const walletBalances = accountAssetsBalances[chain]?.wallet ?? {};
  const balance = getBalance(walletBalances, contractAddress);

  return (
    <Container>
      <HeaderBlock
        centerItems={[
          {
            custom: <TokenIcon url={imageUrl} chain={chain} size={24} />,
          },
          {
            title: ` ${assetData.name} ${t('label.on_network', { network: networkName })}`,
          },
        ]}
        centerItemsStyle={[
          { maxWidth: screenWidth * 0.8 },
          networkName.length > 10 && { marginLeft: screenWidth * 0.05 },
        ]}
        customOnBack={() => navigation.goBack()}
        noPaddingTop
      />
      <Content scrollEnabled={!pointerVisible} bounces={false} paddingHorizontal={0} paddingVertical={0}>
        <Container style={{ alignItems: 'center' }}>
          <Spacing h={20} />
          <HeaderContent
            chain={chain}
            period={selectedPeriod}
            tokenRate={pointerVisible ? convertDecimalNumber(pointerTokenValue) : tokenRateCMC || tokenRate}
            marketDetails={marketDetailsQuery}
            tokenDetails={tokenDetailsQuery}
          />
          <Spacing h={10} />

          <AnimatedGraph
            period={selectedPeriod}
            tokenDetailsData={tokenDetailsQuery.data}
            marketData={marketDetailsQuery.data}
            historicData={historicalTokenPricesQuery}
            onChangePointer={(item, isActive) => {
              setPointerTokenValue(item.value);
              setPointerVisible(isActive);
            }}
          />
          <Spacing h={10} />
          <DurationSelection selectedPeriod={selectedPeriod} onSelectPeriod={setSelectedPeriod} />
          <Spacing h={20} />
          <YourBalanceContent
            tokenRate={tokenRateCMC || tokenRate}
            period={selectedPeriod}
            balance={balance}
            marketDetails={marketDetailsQuery}
            assetData={assetData}
          />
          <Spacing h={36} />
          <TokenAnalytics
            tokenRate={tokenRateCMC || tokenRate}
            tokenDetails={tokenDetailsQuery}
            marketDetails={marketDetailsQuery}
          />
          <Spacing h={130} />
        </Container>
      </Content>
      {!isEtherspotAccount(activeAccount) && <AnimatedFloatingActions assetData={assetData} />}
    </Container>
  );
};

export default AssetScreen;
