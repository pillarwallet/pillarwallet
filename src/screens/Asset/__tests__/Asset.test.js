// @flow
/*
    Pillar Wallet: the personal data locker
    Copyright (C) 2019 Stiftung Pillar Project

    This program is free software; you can redistribute it and/or modify
    it under the terms of the GNU General Public License as published by
    the Free Software Foundation; either version 2 of the License, or
    (at your option) any later version.

    This program is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU General Public License for more details.

    You should have received a copy of the GNU General Public License along
    with this program; if not, write to the Free Software Foundation, Inc.,
    51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
*/
import * as React from 'react';
import renderer from 'react-test-renderer';
import Asset from 'screens/Asset';
import thunk from 'redux-thunk';
import configureMockStore from 'redux-mock-store';
import { Provider } from 'react-redux';
import { ThemeProvider } from 'styled-components/native';
import { I18nextProvider } from 'react-i18next';
import i18n from 'services/localisation/testing';

import { initialState as smartWalletState } from 'reducers/smartWalletReducer';
import { initialState as balancesState } from 'reducers/balancesReducer';
import { initialState as paymentNetworkState } from 'reducers/paymentNetworkReducer';
import { initialState as userState } from 'reducers/userReducer';
import { initialState as historyState } from 'reducers/historyReducer';
import { initialState as assetsState } from 'reducers/assetsReducer';
import { initialState as ratesState } from 'reducers/ratesReducer';
import { initialState as accountsState } from 'reducers/accountsReducer';
import { initialState as appSettingsState } from 'reducers/appSettingsReducer';
import { initialState as userSettingsState } from 'reducers/userSettingsReducer';
import { initialState as exchangeState } from 'reducers/exchangeReducer';
import { initialState as referralsState } from 'reducers/referralsReducer';

import { defaultTheme } from 'utils/themes';


const mockStore = configureMockStore([thunk]);

const initialStore = mockStore({
  user: userState,
  smartWallet: smartWalletState,
  balances: balancesState,
  paymentNetwork: paymentNetworkState,
  history: historyState,
  assets: assetsState,
  rates: ratesState,
  accounts: accountsState,
  appSettings: appSettingsState,
  userSettings: userSettingsState,
  exchange: exchangeState,
  referrals: referralsState,
});


const Component = (store, navigation) => (
  renderer.create(
    <ThemeProvider theme={defaultTheme}>
      <Provider store={store}>
        <I18nextProvider i18n={i18n}>
          <Asset navigation={navigation} />
        </I18nextProvider>
      </Provider>
    </ThemeProvider>)
);

describe('Asset', () => {
  const navigation = {
    addListener: jest.fn(),
    state: {
      params: {
        assetData: {
          token: 'PLR',
          patternIcon: 'http://icons/plr?size=3',
        },
      },
    },
  };

  beforeEach(() => {
    jest.useFakeTimers();
  });

  it('renders the Asset Screen correctly', () => {
    const component = Component(initialStore, navigation).toJSON();
    expect(component).toMatchSnapshot();
  });
});
