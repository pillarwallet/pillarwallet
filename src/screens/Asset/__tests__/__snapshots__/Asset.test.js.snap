// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Asset renders the Asset Screen correctly 1`] = `
<View
  style={
    Array [
      Object {
        "backgroundColor": "#FAFAFA",
        "flexBasis": 0,
        "flexGrow": 1,
        "flexShrink": 1,
      },
    ]
  }
>
  <View
    style={
      Object {
        "backgroundColor": "#FAFAFA",
        "width": "100%",
      }
    }
  >
    <View
      style={
        Array [
          Object {
            "height": 58,
            "paddingHorizontal": 16,
            "paddingVertical": 10,
            "width": "100%",
          },
        ]
      }
    >
      <View
        style={
          Array [
            Object {
              "alignItems": "center",
              "flexDirection": "row",
              "justifyContent": "space-between",
              "width": "100%",
            },
          ]
        }
      >
        <View
          style={
            Array [
              Object {
                "alignItems": "center",
                "flexBasis": 0,
                "flexDirection": "row",
                "flexGrow": 1,
                "flexShrink": 1,
                "height": 36,
                "justifyContent": "flex-start",
              },
              Object {},
            ]
          }
        >
          <View
            accessible={true}
            focusable={true}
            onClick={[Function]}
            onResponderGrant={[Function]}
            onResponderMove={[Function]}
            onResponderRelease={[Function]}
            onResponderTerminate={[Function]}
            onResponderTerminationRequest={[Function]}
            onStartShouldSetResponder={[Function]}
            style={
              Object {
                "alignItems": "center",
                "backgroundColor": "#EBF0F6",
                "borderRadius": 36,
                "height": 36,
                "justifyContent": "center",
                "opacity": 1,
                "paddingBottom": 0,
                "paddingLeft": 0,
                "paddingRight": 0,
                "paddingTop": 0,
                "position": "relative",
                "width": 36,
              }
            }
          >
            <View
              name="back"
              style={
                Array [
                  Object {
                    "color": "#0A1427",
                    "fontSize": 18,
                    "marginBottom": undefined,
                    "marginLeft": undefined,
                    "marginRight": undefined,
                    "marginTop": undefined,
                    "paddingTop": 0,
                  },
                  Object {},
                ]
              }
            />
          </View>
        </View>
        <View
          style={
            Array [
              Object {
                "alignItems": "center",
                "flexBasis": 0,
                "flexDirection": "row",
                "flexGrow": 4,
                "flexShrink": 1,
                "height": 40,
                "justifyContent": "center",
                "paddingBottom": 0,
                "paddingLeft": 12,
                "paddingRight": 12,
                "paddingTop": 0,
              },
            ]
          }
        />
        <View
          style={
            Array [
              Object {
                "alignContent": "center",
                "alignItems": "center",
                "flexBasis": 0,
                "flexDirection": "row",
                "flexGrow": 1,
                "flexShrink": 1,
                "flexWrap": "wrap",
                "height": 36,
                "justifyContent": "flex-end",
              },
            ]
          }
        />
      </View>
    </View>
  </View>
  <View
    onLayout={[Function]}
    style={
      Array [
        Object {
          "flex": 1,
        },
        Object {
          "paddingBottom": 0,
        },
      ]
    }
  >
    <RCTScrollView
      contentContainerStyle={
        Array [
          Object {
            "flexGrow": 1,
          },
          Object {
            "paddingBottom": 120,
          },
        ]
      }
      contentInsetAdjustmentBehavior="automatic"
      keyboardShouldPersistTaps="handled"
      refreshControl={
        <RefreshControl
          onRefresh={[Function]}
          refreshing={false}
        />
      }
      scrollEventThrottle={0}
    >
      <RCTRefreshControl />
      <View>
        <View
          style={
            Array [
              Object {
                "height": 250,
                "justifyContent": "center",
                "width": "100%",
              },
            ]
          }
        >
          <View
            style={
              Array [
                Object {
                  "height": 220,
                  "width": "100%",
                },
              ]
            }
          >
            <View
              addShadow={false}
              borderRadius={54}
              diameter={108}
              elevation={6}
              isListed={true}
              left={375}
              shadowOpacity={0.15}
              style={
                Array [
                  Object {
                    "alignItems": "center",
                    "backgroundColor": "#FFFFFF",
                    "borderColor": "#FFFFFF",
                    "borderStyle": "solid",
                    "borderWidth": 2,
                    "elevation": 0,
                    "height": 108,
                    "justifyContent": "center",
                    "left": 375,
                    "position": "absolute",
                    "shadowColor": "#0A1427",
                    "shadowOffset": Object {
                      "height": 0,
                      "width": 0,
                    },
                    "shadowOpacity": 0,
                    "shadowRadius": 6,
                    "top": 125.5,
                    "width": 108,
                    "zIndex": 3,
                  },
                  Object {
                    "transform": Array [
                      Object {
                        "translateX": -54,
                      },
                      Object {
                        "translateY": -54,
                      },
                    ],
                  },
                ]
              }
              top={125.5}
              zIndex={3}
            >
              <CMIFColorMatrixImageFilter
                matrix={
                  Array [
                    1,
                    0,
                    0,
                    0,
                    0,
                    0,
                    1,
                    0,
                    0,
                    0,
                    0,
                    0,
                    1,
                    0,
                    0,
                    0,
                    0,
                    0,
                    1,
                    0,
                  ]
                }
                style={
                  Array [
                    Object {
                      "backgroundColor": "#fff0",
                    },
                    undefined,
                  ]
                }
              />
            </View>
            <View
              addShadow={false}
              borderRadius={35}
              diameter={70}
              elevation={0}
              isListed={true}
              left={266.2}
              shadowOpacity={0}
              style={
                Array [
                  Object {
                    "alignItems": "center",
                    "backgroundColor": "#FFFFFF",
                    "borderColor": "#FFFFFF",
                    "borderStyle": "solid",
                    "borderWidth": 2,
                    "elevation": 0,
                    "height": 70,
                    "justifyContent": "center",
                    "left": 266.2,
                    "position": "absolute",
                    "shadowColor": "#0A1427",
                    "shadowOffset": Object {
                      "height": 0,
                      "width": 0,
                    },
                    "shadowOpacity": 0,
                    "shadowRadius": 6,
                    "top": 65.5,
                    "width": 70,
                    "zIndex": 1,
                  },
                  Object {
                    "transform": Array [
                      Object {
                        "translateX": -35,
                      },
                      Object {
                        "translateY": -35,
                      },
                    ],
                  },
                ]
              }
              top={65.5}
              zIndex={1}
            >
              <CMIFColorMatrixImageFilter
                matrix={
                  Array [
                    0.6065,
                    0.3575,
                    0.036,
                    0,
                    0,
                    0.1065,
                    0.8574999999999999,
                    0.036,
                    0,
                    0,
                    0.1065,
                    0.3575,
                    0.536,
                    0,
                    0,
                    0,
                    0,
                    0,
                    1,
                    0,
                  ]
                }
                style={
                  Array [
                    Object {
                      "backgroundColor": "#fff0",
                    },
                    undefined,
                  ]
                }
              />
            </View>
            <View
              addShadow={false}
              borderRadius={45}
              diameter={90}
              elevation={4}
              isListed={true}
              left={310.33333333333337}
              shadowOpacity={0.05}
              style={
                Array [
                  Object {
                    "alignItems": "center",
                    "backgroundColor": "#FFFFFF",
                    "borderColor": "#FFFFFF",
                    "borderStyle": "solid",
                    "borderWidth": 2,
                    "elevation": 0,
                    "height": 90,
                    "justifyContent": "center",
                    "left": 310.33333333333337,
                    "position": "absolute",
                    "shadowColor": "#0A1427",
                    "shadowOffset": Object {
                      "height": 0,
                      "width": 0,
                    },
                    "shadowOpacity": 0,
                    "shadowRadius": 6,
                    "top": 104.5,
                    "width": 90,
                    "zIndex": 2,
                  },
                  Object {
                    "transform": Array [
                      Object {
                        "translateX": -45,
                      },
                      Object {
                        "translateY": -45,
                      },
                    ],
                  },
                ]
              }
              top={104.5}
              zIndex={2}
            >
              <CMIFColorMatrixImageFilter
                matrix={
                  Array [
                    0.7638999999999999,
                    0.21450000000000002,
                    0.0216,
                    0,
                    0,
                    0.06390000000000001,
                    0.9145,
                    0.0216,
                    0,
                    0,
                    0.06390000000000001,
                    0.21450000000000002,
                    0.7215999999999999,
                    0,
                    0,
                    0,
                    0,
                    0,
                    1,
                    0,
                  ]
                }
                style={
                  Array [
                    Object {
                      "backgroundColor": "#fff0",
                    },
                    undefined,
                  ]
                }
              />
            </View>
            <View
              addShadow={false}
              borderRadius={45}
              diameter={90}
              elevation={4}
              isListed={true}
              left={439.66666666666663}
              shadowOpacity={0.05}
              style={
                Array [
                  Object {
                    "alignItems": "center",
                    "backgroundColor": "#FFFFFF",
                    "borderColor": "#FFFFFF",
                    "borderStyle": "solid",
                    "borderWidth": 2,
                    "elevation": 0,
                    "height": 90,
                    "justifyContent": "center",
                    "left": 439.66666666666663,
                    "position": "absolute",
                    "shadowColor": "#0A1427",
                    "shadowOffset": Object {
                      "height": 0,
                      "width": 0,
                    },
                    "shadowOpacity": 0,
                    "shadowRadius": 6,
                    "top": 104.5,
                    "width": 90,
                    "zIndex": 2,
                  },
                  Object {
                    "transform": Array [
                      Object {
                        "translateX": -45,
                      },
                      Object {
                        "translateY": -45,
                      },
                    ],
                  },
                ]
              }
              top={104.5}
              zIndex={2}
            >
              <CMIFColorMatrixImageFilter
                matrix={
                  Array [
                    0.7638999999999999,
                    0.21450000000000002,
                    0.0216,
                    0,
                    0,
                    0.06390000000000001,
                    0.9145,
                    0.0216,
                    0,
                    0,
                    0.06390000000000001,
                    0.21450000000000002,
                    0.7215999999999999,
                    0,
                    0,
                    0,
                    0,
                    0,
                    1,
                    0,
                  ]
                }
                style={
                  Array [
                    Object {
                      "backgroundColor": "#fff0",
                    },
                    undefined,
                  ]
                }
              />
            </View>
            <View
              addShadow={false}
              borderRadius={35}
              diameter={70}
              elevation={0}
              isListed={true}
              left={483.8}
              shadowOpacity={0}
              style={
                Array [
                  Object {
                    "alignItems": "center",
                    "backgroundColor": "#FFFFFF",
                    "borderColor": "#FFFFFF",
                    "borderStyle": "solid",
                    "borderWidth": 2,
                    "elevation": 0,
                    "height": 70,
                    "justifyContent": "center",
                    "left": 483.8,
                    "position": "absolute",
                    "shadowColor": "#0A1427",
                    "shadowOffset": Object {
                      "height": 0,
                      "width": 0,
                    },
                    "shadowOpacity": 0,
                    "shadowRadius": 6,
                    "top": 65.5,
                    "width": 70,
                    "zIndex": 1,
                  },
                  Object {
                    "transform": Array [
                      Object {
                        "translateX": -35,
                      },
                      Object {
                        "translateY": -35,
                      },
                    ],
                  },
                ]
              }
              top={65.5}
              zIndex={1}
            >
              <CMIFColorMatrixImageFilter
                matrix={
                  Array [
                    0.6065,
                    0.3575,
                    0.036,
                    0,
                    0,
                    0.1065,
                    0.8574999999999999,
                    0.036,
                    0,
                    0,
                    0.1065,
                    0.3575,
                    0.536,
                    0,
                    0,
                    0,
                    0,
                    0,
                    1,
                    0,
                  ]
                }
                style={
                  Array [
                    Object {
                      "backgroundColor": "#fff0",
                    },
                    undefined,
                  ]
                }
              />
            </View>
          </View>
        </View>
        <View
          style={
            Array [
              Object {
                "alignItems": "center",
                "justifyContent": "center",
                "marginBottom": 20,
                "marginLeft": 20,
                "marginRight": 20,
                "marginTop": 0,
                "paddingBottom": 8,
              },
            ]
          }
        >
          <View
            style={
              Array [
                Object {
                  "alignItems": "center",
                  "flexDirection": "row",
                  "justifyContent": "center",
                },
              ]
            }
          >
            <Text
              style={
                Array [
                  Object {
                    "color": "#0A1427",
                    "fontFamily": "EuclidCircularB-Medium",
                    "fontSize": 36,
                    "lineHeight": 54,
                    "textAlign": "center",
                    "textAlignVertical": "center",
                  },
                ]
              }
            >
              tokenValue
            </Text>
          </View>
          <View
            style={
              Array [
                Object {
                  "flexDirection": "row",
                },
              ]
            }
          >
            <Text
              style={
                Array [
                  Object {
                    "color": "#0A1427",
                    "fontFamily": "EuclidCircularB-Regular",
                    "fontSize": 12,
                    "lineHeight": 18,
                    "textAlign": "center",
                    "textAlignVertical": "center",
                  },
                ]
              }
            >
              £0
            </Text>
          </View>
        </View>
      </View>
    </RCTScrollView>
  </View>
  <View
    style={
      Array [
        Object {
          "alignItems": "center",
          "backgroundColor": "#FFFFFF",
          "borderRadius": 20,
          "elevation": 6,
          "flexDirection": "row",
          "paddingHorizontal": 10,
          "shadowColor": "#000",
          "shadowOffset": Object {
            "height": 8,
            "width": 0,
          },
          "shadowOpacity": 0.05,
          "shadowRadius": 16,
        },
      ]
    }
  >
    <View
      accessible={true}
      focusable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "alignItems": "center",
          "opacity": 1,
          "paddingBottom": 12,
          "paddingHorizontal": 4,
          "paddingTop": 16,
        }
      }
      testID="FloatingButtonItem"
    >
      <View
        style={
          Array [
            Object {
              "alignItems": "center",
              "flexBasis": 0,
              "flexGrow": 1,
              "flexShrink": 1,
              "justifyContent": "center",
              "marginHorizontal": 24,
            },
          ]
        }
      />
      <Text
        style={
          Array [
            Object {
              "fontFamily": "EuclidCircularB-Regular",
              "textAlignVertical": "center",
            },
            Object {
              "color": "#0A1427",
            },
            Object {
              "fontSize": 14,
              "lineHeight": 22,
            },
            false,
            Array [
              Object {
                "marginTop": 6,
                "textAlign": "center",
              },
            ],
          ]
        }
      >
        button.swap
      </Text>
    </View>
    <View
      accessible={true}
      focusable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "alignItems": "center",
          "opacity": 1,
          "paddingBottom": 12,
          "paddingHorizontal": 4,
          "paddingTop": 16,
        }
      }
      testID="FloatingButtonItem"
    >
      <View
        style={
          Array [
            Object {
              "alignItems": "center",
              "flexBasis": 0,
              "flexGrow": 1,
              "flexShrink": 1,
              "justifyContent": "center",
              "marginHorizontal": 24,
            },
          ]
        }
      />
      <Text
        style={
          Array [
            Object {
              "fontFamily": "EuclidCircularB-Regular",
              "textAlignVertical": "center",
            },
            Object {
              "color": "#0A1427",
            },
            Object {
              "fontSize": 14,
              "lineHeight": 22,
            },
            false,
            Array [
              Object {
                "marginTop": 6,
                "textAlign": "center",
              },
            ],
          ]
        }
      >
        button.send
      </Text>
    </View>
  </View>
</View>
`;
